File: parser.py
class Tree:
    def __init__(self,indexes,data):
        self.open = indexes[0]
        self.close = indexes[1]
        self.data = data
        self.childs = []

    @staticmethod
    def addNode(tree,indexes,data):
        last = None
        if len(tree.childs) > 0 :
            last = tree.childs[-1]
        else:
            tree.childs.append(Tree(indexes,data))
        if last:
            if last.close == indexes[0] - 1:
                tree.childs.append(Tree(indexes,data))
                return
            else:
                Tree.addNode(tree.childs[-1],indexes,data)

    def display(self, index):
        if self:
            print("\t" * index, end = "")
            print(self.data[0].strip('<').strip('>'), end = "")
            if len(self.data) > 1:
                print(" : " + self.data[1])
            else:
                print()
            if self.childs:
                for child in self.childs:
                    child.display(index + 1)
        else:
            return


class File:
    def __init__(self,file_name):
        self.file = open(file_name,'r')

    def getlines(self):
        return self.file.read().split('\n')

    def __del__(self):
        self.file.close()


class Parser:
    def __init__(self,file_name):
        self.file = File(file_name)
        self.lines = self.file.getlines()
        self.XMLconfiguration = {}
        self.elements = []
        self.xmltree = None

    def findclosetag(self,elements,starting_position, tag):
        for index, element in enumerate(elements[starting_position:]):
            for subelement in element:
                if index > 0:
                    if "<" + tag + ">" in subelement:
                        return False
                if "</" + tag + ">" in subelement:
                    return True, starting_position, index + starting_position
        return False

    def getXMLconfiguration(self):
        configuration = self.lines[0].split(' ')
        configuration[-1] = configuration[-1][:-2]
        for element in configuration[1:]:
            elements = element.split("=")
            self.XMLconfiguration.update({elements[0] : elements[1]})
        self.lines = self.lines[1:]


    def elementSegmentation(self):
        for index, line in enumerate(self.lines):
            first_tag = line[:line.find('>') + 1].strip()
            tag = line[line.find('>') + 1:].split("<")
            if len(tag) > 1:
                tag[-1] = "<" + tag[-1]
                self.elements.append([first_tag,tag[0],tag[1]])
            elif len(first_tag) > 1:
                self.elements.append([first_tag])

    def buildXMLTree(self):
        for index, element in enumerate(self.elements):
            if len(element) == 1:
                attribute = element[0].strip('<').strip('>')
                if "/" not in attribute:
                    validitate =  self.findclosetag(self.elements,index,attribute)
                    if validitate[1] == 0:
                        self.xmltree = Tree(validitate[1:],element)
                    else:
                        Tree.addNode(self.xmltree,validitate[1:],element)
            elif len(element) > 1:
                attribute = element[0].strip('<').strip('>')
                validitate = self.findclosetag(self.elements,index,attribute)
                Tree.addNode(self.xmltree,validitate[1:],element)


if __name__ == "__main__":
    parser = Parser("licenta.xml")
    parser.getXMLconfiguration()
    parser.elementSegmentation()
    parser.buildXMLTree()
    print(parser.XMLconfiguration)
    parser.xmltree.display(0)

File: c++/parser.cpp
#include "parser.h"


XMLFile::XMLFile(const std::string & title) {
	file = new std::ifstream(title);
}

std::vector<std::string> XMLFile::getlines() {
	std::vector<std::string> lines{};
	std::string line;
	while (std::getline(*this->file, line)) {
		lines.push_back(line);
	}
	return lines;
}

XMLFile::~XMLFile() {
	if (file != nullptr) {
		delete file;
	}
}

/* --------------------------------------------------------------------------------------- */

Tree::Tree(std::pair<uint16_t, uint16_t> indexes, std::vector<std::string> data) {
	this->open = indexes.first;
	this->close = indexes.second;
	this->data = data;
	this->childs = {};
}

void Tree::addNode(Tree &tree, std::pair<uint16_t, uint16_t> indexes, std::vector<std::string> data) {
	Tree *last = nullptr;
	if (tree.childs.size() > 0) {
		last = tree.childs.back();
	}
	else {
		tree.childs.push_back(new Tree(indexes, data));
	}
	if (last != nullptr) {
		if (last->close == indexes.first - 1) {
			tree.childs.push_back(new Tree(indexes, data));
			return;
		}
		else {
			Tree::addNode(*tree.childs.back(), indexes, data);
		}
	}
}

void Tree::display(uint8_t index) {
	if (this != nullptr) {
		std::cout << std::string(index, '\t');
		std::cout << this->data.front();
		if (this->data.size() > 1) {
			std::cout << " : " << this->data[1] << std::endl;
		}
		else {
			std::cout << std::endl;
		}
	}
	else {
		return;
	}
}

/* --------------------------------------------------------------------------------------- */

void FormatOperations::displayMap(Parser parser){
	for (const auto & element : parser.XMLconfiguration) {
		std::cout << element.first << " : " << element.second << std::endl;
	}
	std::cout << std::endl;
}

std::vector<std::string> FormatOperations::split(std::string string, std::string delimiter) {
	std::vector<std::string> tokens = {};
	size_t pos = 0;
	std::string token;
	while ((pos = string.find(delimiter)) != std::string::npos) {
		token = string.substr(0, pos);
		//std::cout << token << std::endl;
		tokens.push_back(token);
		string.erase(0, pos + delimiter.length());
	}
	if (string.length() > 0) {
		tokens.push_back(string);
	}
	return tokens;
}

std::vector<std::string> FormatOperations::substrings(std::vector<std::string> target, uint16_t index) {
	std::vector<std::string> substr = {};
	for (std::vector<std::string>::iterator it = target.begin() + index; it != target.end(); it++) {
		substr.push_back(*it);
	}
	return substr;
}

void FormatOperations::strip(std::string & target, std::string character) {
	int index_bracket = target.find(character);
	target.erase(index_bracket, character.length());
}

void FormatOperations::displayVectors(Parser parser) {
	for (const auto & element : parser.elements) {
		for (const auto & subelement : element) {
			std::cout << subelement;
		}
		std::cout << std::endl;
	}
}

/* --------------------------------------------------------------------------------------- */

std::pair<uint16_t, uint16_t> Parser::findclosetag(std::vector<std::vector<std::string>> elements,
	uint16_t starting_position, std::string tag) {
	for (std::vector<std::vector<std::string>>::iterator element = elements.begin() + starting_position; element != elements.end(); element++) {
		for (const auto & subelement : (*element)) {
			if (element - elements.begin() > 0) {
				std::string substring = std::string("<" + tag + ">");
				if (subelement.find(substring) != std::string::npos) {
					return std::pair<uint16_t, uint16_t>{-1, -1};
				}
			}
			std::string substring = std::string("</" + tag + ">");
			if (subelement.find(substring) != std::string::npos) {
				return std::pair<uint16_t, uint16_t>{starting_position, element - elements.begin() + starting_position};
			}
		}

	}
	return std::pair<uint16_t, uint16_t>{-1, -1};
}

Parser::Parser(const std::string & filename) {
	this->file = new XMLFile(filename);
	this->lines = this->file->getlines();
	this->XMLconfiguration = {};
	this->elements = {};
}

std::map<std::string, std::string> Parser::getconfig() {
		return this->XMLconfiguration;
	}

void Parser::getXMLConfiguration() {
	std::vector<std::string> configuration = FormatOperations::split(this->lines[0], std::string(" "));
	for (const auto & element : configuration) {
		if (element != configuration.front()) {
			std::vector<std::string> elements = FormatOperations::split(element, std::string("="));
			this->XMLconfiguration.insert(std::pair<std::string, std::string>{elements[0], elements[1]});
		}
	}
	this->lines = FormatOperations::substrings(this->lines, 1);
}

void Parser::elementSegmentation() {
	for (std::vector<std::string>::iterator line = this->lines.begin(); line != this->lines.end(); line++) {
		unsigned short int bracket_position = (*line).find(std::string(">")) + 1;
		std::string line_copy(*line);
		std::string first_tag = line_copy.erase(bracket_position, line_copy.length() - bracket_position);
		std::string tag = (*line).erase(0, bracket_position);
		std::vector<std::string> tags = FormatOperations::split(tag, std::string("<"));
		if (tags.size() > 1) {
			std::string aux = "<" + tags.back();
			tags.pop_back();
			tags.push_back(aux);
			this->elements.push_back(std::vector<std::string>{ first_tag,tags[0],tags[1] });
		}
		else if (first_tag.length() > 1) {
			this->elements.push_back(std::vector<std::string>{first_tag});
		}
	}
}

void Parser::buildXMLTree() {
	for (std::vector<std::vector<std::string>>::iterator line = this->elements.begin(); line != this->elements.end(); line++) {
		if ((*line).size() == 1) {
			std::string attribute((*line)[0]);
			FormatOperations::strip(attribute, std::string("<"));
			FormatOperations::strip(attribute, std::string(">"));
			if (attribute.find(std::string("/")) > attribute.length()) {
				std::pair<uint16_t, uint16_t> validitate = this->findclosetag(this->elements, (uint16_t)(line - this->elements.begin()), attribute);
				if (validitate.first == 0) {
					this->xmltree = new Tree(validitate, (*line));
				}
				else {
					Tree::addNode(*this->xmltree, validitate, *line);
				}
			}
		}
		else if ((*line).size() > 1) {
			std::string attribute((*line)[0]);
			FormatOperations::strip(attribute, std::string("<"));
			FormatOperations::strip(attribute, std::string(">"));
			std::pair<uint16_t, uint16_t> validitate = this->findclosetag(this->elements, (uint16_t)(line - this->elements.begin()), attribute);
			Tree::addNode(*this->xmltree, validitate, *line);
		}
	}
}

Tree Parser::getTree() {
	return *this->xmltree;
}

/* --------------------------------------------------------------------------------------- */


int main() {
	Parser parser("licenta.xml");
	parser.getXMLConfiguration();
	std::map<std::string, std::string> xmlconfiguration = parser.getconfig();
	FormatOperations::displayMap(parser);
	parser.elementSegmentation();
	parser.buildXMLTree();
	parser.getTree().display(0);
	system("pause");
	return 0;
}